mixin/component-js@1/mixin.php

```diff
@@ -0,0 +1,217 @@
+    ]);
+  }
+
+  /**
+   * @internal
+   * @see \CRM_Utils_Hook::alterBundle()
+   */
+  public function alterBundle(\CRM_Core_Resources_Bundle $bundle) {
+    if ($bundle->name === 'coreResources') {
+      $url = Civi::service('asset_builder')->getUrl(static::ASSET, ['id' => $this->getId()]);
+      $bundle->addModuleUrl($url);
+    }
+  }
+
+  /**
+   * @internal
+   * @return array
+   */
+  public function getStaticPaths(): array {
+    $paths = [];
+    $event = Civi\Core\Event\GenericHookEvent::create(['path' => &$paths]);
+    Civi::dispatcher()->dispatch('hook_civicrm_componentJsPaths', $event);
+    return $paths;
+  }
+
+  /**
+   * @return string
+   */
+  public function getId(): string {
+    $id = Civi::cache('long')->get('component-js-id');
+    if ($id === NULL) {
+      $id = CRM_Utils_String::createRandom(8, CRM_Utils_String::ALPHANUMERIC);
+      Civi::cache('long')->set('component-js-id', $id, 7 * 24 * 60 * 60);
+    }
+    return $id;
+  }
+
+  public function flush(): void {
+    Civi::cache('long')->delete('component-js-id');
+  }
+
+  /**
+   * @return string
+   */
+  private function getTemplate(): string {
+    $fp = fopen(__FILE__, 'r');
+    fseek($fp, __COMPILER_HALT_OFFSET__);
+    $template = stream_get_contents($fp);
+    fclose($fp);
+    return $template;
+  }
+
+}
+
+/**
+ * As a mixin, we receive a notification for each extension that enables `component-js@1`.
+ *
+ * @param \CRM_Extension_MixInfo $mixInfo
+ * @param \CRM_Extension_BootCache $bootCache
+ */
+return function ($mixInfo, $bootCache) {
+
+  ComponentJs::instance()->register();
+
+  /**
+   * Register this extension with the ESM import-map ('my-extension/' => '/var/www/sites/default/civicrm/ext/my-extension').
+   *
+   * @see \CRM_Utils_Hook::esmImportMap()
+   */
+  Civi::dispatcher()->addListener('&hook_civicrm_esmImportMap', function(\Civi\Esm\ImportMap $importMap) use ($mixInfo) {
+    if ($mixInfo->isActive()) {
+      $importMap->addPrefix($mixInfo->longName . '/', $mixInfo->longName);
+    }
+  }, 500);
+
+  /**
+   * Statically register each *.js or *.mjs file from $EXTENSION/js/component/.
+   */
+  Civi::dispatcher()->addListener('&hook_civicrm_componentJsPaths', function(array &$paths) use ($mixInfo) {
+    if ($mixInfo->isActive()) {
+      $files = array_merge(
+        (array) glob($mixInfo->getPath(ComponentJs::SUBDIR . '/*.js')),
+        (array) glob($mixInfo->getPath(ComponentJs::SUBDIR . '/*.mjs')),
+      );
+
+      foreach ($files as $file) {
+        $tag = preg_replace('/\.(js|mjs)$/', '', basename($file));
+        $paths[$tag] = $mixInfo->longName . '/' . ComponentJs::SUBDIR . '/' . basename($file) . '?ts=' . filemtime($file);
+      }
+    }
+  });
+
+};
+
+###############################################################################
+## Below, we stop executing PHP. The rest of the file contains the template
+## for "component.js".
```

Comment:
why can't this be in a separate file?

Reply:
Mixin backports/versioning are generally based on one-file/one-mixin.

From POV of civicrm-core and PHP interpreter, I suppose one could do more files. But it’s definitely messier on civix side. I’d probably want to have a few more needs before biting off that task